core/lib/iq-seeder.c

```diff
@@ -0,0 +1,246 @@
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the Institute nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE INSTITUTE AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE INSTITUTE OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ * This file is part of the Contiki operating system.
+ *
+ */
+
+/**
+ * \file
+ *         Self-seeds with I/Q LSBs.
+ * \author
+ *         Konrad Krentz <konrad.krentz@gmail.com>
+ */
+
+#include "lib/iq-seeder.h"
+#include "net/netstack.h"
+#include "lib/aes-128.h"
+#include <string.h>
+
+#ifdef IQ_SEEDER_CONF_WITH_CHANNEL_HOPPING
+#define WITH_CHANNEL_HOPPING IQ_SEEDER_CONF_WITH_CHANNEL_HOPPING
+#else /* IQ_SEEDER_CONF_WITH_CHANNEL_HOPPING */
+#define WITH_CHANNEL_HOPPING 1
+#endif /* IQ_SEEDER_CONF_WITH_CHANNEL_HOPPING */
+
+#ifdef IQ_SEEDER_CONF_SAMPLING_RATE
+#define SAMPLING_RATE IQ_SEEDER_CONF_SAMPLING_RATE
+#else /* IQ_SEEDER_CONF_SAMPLING_RATE */
+#define SAMPLING_RATE 200
+#endif /* IQ_SEEDER_CONF_SAMPLING_RATE */
+
+#define COLUMN_COUNT  50
+#define ROW_COUNT     16
+
+#define DEBUG 0
+#if DEBUG
+#include <stdio.h>
+#define PRINTF(...) printf(__VA_ARGS__)
+#else /* DEBUG */
+#define PRINTF(...)
+#endif /* DEBUG */
+
+/* extractor seed originates from RANDOM.ORG */
+static const uint8_t toeplitz[COLUMN_COUNT + ROW_COUNT - 1] =
+    {  93 ,  50 , 210 , 134 ,  79 ,  52 , 237 , 192 ,  40 , 201 ,
+        3 , 184 , 152 ,  74 ,  27 ,  28 ,  32 , 111 ,  79 , 222 ,
+      174 ,  51 , 223 ,  66 , 152 , 211 , 234 , 124 ,  92 ,  64 ,
+      206 , 169 , 227 , 155 , 106 ,  87 , 207 , 135 , 238 , 101 ,
+      254 , 163 ,  55 ,  76 ,  50 ,  40 ,   4 , 149 ,  27 ,   1 ,
+      127 , 159 , 160 ,  91 , 251 , 179 , 186 , 200 , 225 ,  47 ,
+      235 , 223 ,  39 , 117 ,  19 };
+
+/*---------------------------------------------------------------------------*/
+static uint8_t
+get_toeplitz_element(uint8_t row, uint8_t column)
+{
+  uint8_t min;
+
+  min = row < column ? row : column;
+  row -= min;
+  column -= min;
+
+  return toeplitz[row ? COLUMN_COUNT - 1 + row : column];
+}
+/*---------------------------------------------------------------------------*/
+#if 0
```

Comment:
I would prefer either to remove the commented block, or add some comments about why is disabled